saw-core/src/Verifier/SAW/Term/Functor.hs

```diff
@@ -382,8 +383,8 @@ alphaEquiv = term
     term (Unshared tf1) (Unshared tf2) = termf tf1 tf2
     term (Unshared tf1) (STApp{stAppTermF = tf2}) = termf tf1 tf2
     term (STApp{stAppTermF = tf1}) (Unshared tf2) = termf tf1 tf2
-    term (STApp{stAppIndex = i1, stAppTermF = tf1})
-         (STApp{stAppIndex = i2, stAppTermF = tf2}) = i1 == i2 || termf tf1 tf2
+    term (STApp{stAppIndex = i1, stAppHash = h1, stAppTermF = tf1})
```

Comment:
I don't think this change is correct. The `stAppHash` field is just caching the hash result; it shouldn't be used for equality comparison.

Reply:
I made this change because the [contract of `hashWithSalt`](https://hackage.haskell.org/package/hashable-1.4.2.0/docs/Data-Hashable.html#v:hashWithSalt) dictates that if two terms compare equally according to `==`, they must compare equally according to `hashWithSalt`. Without this check, and with these changes to `Hashable`, it is possible to violate that contract, if two terms have the same index but different hashes. I think this is unlikely, if terms are always constructed in a way that assigns indices appropriately, but it is still possible. I also think that I encountered some surprising compilation and/or test failures without this check in place, but I don't remember with certainty, so I'll see if I can recreate the issue(s).

Reply:
As it turns out, restoring the original behavior of `Eq` does _not_ seem to produce compilation or test failures. Nevertheless, I'm a little leery of violating the hashing contract. Do you have any situations in mind in which this modified `Eq` instance is likely to produce incorrect or incoherent comparisons?

Reply:
`stAppIndex` is unique, thus `i1 == i2` implies that `h1 == h2`.
Separately, it's possible to make the check faster
```
i1 == i2 || (h1 == h2 && termf tf1 tf2)
```
but this would require to manually derive the `Hashable (TermF e)` in order to exclude `LocalName` from the computation for `Lambda` and `Pi`. I leave it up to you, @samcowger

Reply:
Alright, I've reverted my change to `==`, and left a comment on my thought process around it.